lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -144,54 +181,79 @@ def _on_index_requested(self, event: IndexRequestedEvent) -> None:
-        self.update_endpoints(event.relation)
         self.opensearch_provides.set_version(rel_id, self.opensearch.version)
+        self.opensearch_provides.set_credentials(rel_id, username, pwd)
         self.update_certs(rel_id)
+        self.update_endpoints(event.relation)
 
     def create_opensearch_users(
-        self, username: str, hashed_pwd: str, access_control: Dict[str, List[str]]
+        self,
+        username: str,
+        hashed_pwd: str,
+        index: str,
+        extra_user_roles: str,
     ):
         """Creates necessary opensearch users and permissions for this relation.
 
         Args:
             username: Username to be created
             hashed_pwd: the hash of the password to be assigned to the user
-            access_control: A dict of roles, permissions, and action groups to be applied to the
-                user. A new role will be created to contain the requested permissions and action
-                groups.
+            index: the index to which the users must be granted access
+            extra_user_roles: the level of permissions that the user should be given. Can be a
+                comma-separated list of roles, which should result in a merged list of permissions.
 
         Raises:
             OpenSearchUserMgmtError if user creation fails
         """
-        roles = access_control.get("roles")
-        permissions = access_control.get("permissions")
-        action_groups = access_control.get("action_groups")
-
-        if permissions or action_groups:
-            # combine action groups and perms into a new role for this relation.
-            try:
-                self.user_manager.create_role(
-                    role_name=username,
-                    permissions=permissions,
-                    action_groups=action_groups,
-                )
-                roles.append(username)
-            except OpenSearchUserMgmtError as err:
-                logger.error(err)
-                raise
-
         try:
+            # Create a new role for this relation, encapsulating the permissions we care about. We
+            # can't create a "default" and an "admin" role once because the permissions need to be
+            # set to this relation's specific index.
+            self.user_manager.create_role(
+                role_name=username,
+                permissions=self.get_extra_user_role_permissions(extra_user_roles, index),
+            )
+            roles = [username]
             self.user_manager.create_user(username, roles, hashed_pwd)
-            if roles:
-                self.user_manager.patch_user(
-                    username,
-                    [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
-                )
+            self.user_manager.patch_user(
+                username,
+                [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
+            )
```

Comment:
**question** I remember you implemented some rollback logic some time ago, which now seems to have disappeared or was it a different file/feature?

Reply:
I believe it was in opensearch_users.py, the idea now is to allow create_user functions called with the same username to overwrite that user, so if they're called multiple times we know we have the user we expect

Reply:
I believe it was in opensearch_users.py, the idea now is to allow create_user functions called with the same username to overwrite that user, so if they're called multiple times we know we have the user we expect